src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,461 @@
+
+    bool const                               Binary;
+    std::string const                        Norm;
+    bool const                               UseIdf;
+    bool const                               SmoothIdf;
+    bool const                               SublinearTf;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    TfidfVectorizerTransformer(IndexMapType labels, IndexMapType docufreq, std::uint32_t totalnumdocus, bool binary, 
+                               std::string norm, bool use_idf, bool smooth_idf, bool sublinear_tf);
+    TfidfVectorizerTransformer(Archive &ar);
+
+    ~TfidfVectorizerTransformer(void) override = default;
+
+    void save(Archive &ar) const override;
+
+    bool operator==(TfidfVectorizerTransformer const &other) const;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC has problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+
+        typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp> TermFreqMap;
+
+        //todo: will use vector<functor> after string header file is done
+        Components::split_temp( 
+            input, 
+            [&TermFreqMap] (std::string::const_iterator & iter_start, std::string::const_iterator & iter_end) {
+                
+                typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>::iterator iter_tf(TermFreqMap.find(std::make_tuple(iter_start, iter_end)));
+                
+                if (iter_tf != TermFreqMap.end()) {
+                    ++iter_tf->second;
+                } else {
+                    TermFreqMap.insert(std::make_pair(std::make_tuple(iter_start, iter_end), 1));
+                }
+            }
+        );
+
+        std::float_t normVal = 0.0f;
+        std::vector<std::tuple<std::uint32_t, std::float_t>> results;
+        for (auto const & pair : TermFreqMap) {
+            std::string const word = std::string(std::get<0>(pair.first), std::get<1>(pair.first));
+
+            typename IndexMapType::const_iterator const      iter_label(Labels.find(word));
+
+            if (iter_label != Labels.end()) {
+                
+                std::float_t tf;
+                std::float_t idf;
+
+                //calculate tf
+                if (Binary) {
+                    tf = 1.0f;
+                } else if (!SublinearTf) {
+                    tf = static_cast<std::float_t>(pair.second);
+                } else {
+                    tf = 1.0f + static_cast<std::float_t>(log(pair.second));
```

Comment:
>static_cast<std::float_t>(log(pair.second)); [](start = 32, length = 44)

use std::log overload for float and avoid casting. Every time you want to cast, ask yourself if you can avoid it.